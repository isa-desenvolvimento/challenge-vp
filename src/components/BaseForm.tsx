import { Formik, Form } from 'formik'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const BaseForm = ({ initialValues, baseForm, onSubmit }: any) => {
  const renderForm = () => {
    return (
      <Formik
        onSubmit={(data, action) => onSubmit(data, action)}
        initialValues={initialValues}
      >
        {() => <Form>{baseForm}</Form>}
      </Formik>
    )
  }

  return <>{renderForm()}</>
}
