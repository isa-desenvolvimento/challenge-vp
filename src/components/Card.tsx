import {
  StyleCardContainer,
  StyleCardContent,
  StyleCardFooter,
  StyleCardHeader,
} from '@/styles/Card'
import { CardProps } from '@/types/global'

export const Card = ({
  header,
  content,
  footer,
  primary,
  height,
}: CardProps) => {
  const renderHeader = () =>
    header && <StyleCardHeader primary={primary}>{header}</StyleCardHeader>
  const renderContent = () =>
    content && (
      <StyleCardContent primary={primary} height={height}>
        {content}
      </StyleCardContent>
    )
  const renderFooter = () =>
    footer && <StyleCardFooter primary={primary}>{footer}</StyleCardFooter>

  return (
    <StyleCardContainer role="card">
      {renderHeader()}
      {renderContent()}
      {renderFooter()}
    </StyleCardContainer>
  )
}
