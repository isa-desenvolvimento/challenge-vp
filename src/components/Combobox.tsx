import { ComboboxProps } from '@/types/global'

import { DropdownWrapper, StyledSelect, StyledOption } from '@/styles/Combobox'
import { useField } from 'formik'

export const Combobox = ({
  role,
  options,
  placeholder,
  primary,
  onChange,
  name,
  isFormik,
}: ComboboxProps) => {
  const [field, meta] = useField(name)

  const renderSelection = () => {
    if (isFormik) {
      return (
        <StyledSelect primary={primary} {...field} {...meta} role={role}>
          <StyledOption selected value="" name={`${name}.defaultValue`}>
            {placeholder}
          </StyledOption>
          {renderOptions()}
        </StyledSelect>
      )
    }

    return (
      <StyledSelect onChange={onChange} primary={primary}>
        <StyledOption selected value="" name={`${name}.defaultValue`}>
          {placeholder}
        </StyledOption>
        {renderOptions()}
      </StyledSelect>
    )
  }

  const renderOptions = () => {
    return (
      options?.length &&
      options.map(({ value }) => {
        return (
          <StyledOption name={`${name}.${value}`} key={value} value={value}>
            {value}
          </StyledOption>
        )
      })
    )
  }

  return <DropdownWrapper role="combobox">{renderSelection()}</DropdownWrapper>
}
