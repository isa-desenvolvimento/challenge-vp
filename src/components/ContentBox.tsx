import { StyleContainer } from '@/styles/ContentBox'
import { ChildrenProps } from '@/types/global'

export const ContentBox = ({ children }: ChildrenProps) => {
  return <StyleContainer role="contentBox"> {children}</StyleContainer>
}
