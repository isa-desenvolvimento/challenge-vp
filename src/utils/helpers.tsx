import { render } from '@testing-library/react'
import GlobalContext from '../context'

export const renderWithProvider = (children: React.ReactNode) => {
  render(<GlobalContext>{children}</GlobalContext>)
}
