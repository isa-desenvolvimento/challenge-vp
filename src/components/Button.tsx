import { Icon } from '@/components/Icon'
import { StyleButton } from '@/styles/Button'
import { ButtonProps } from '@/types/global'

export const Button = ({
  iconProps,
  primary,
  onClick,
  children,
  size,
  type,
}: ButtonProps) => {
  const renderIcon = () => iconProps && <Icon faIcon={iconProps}></Icon>

  return (
    <StyleButton
      role="button"
      arial-label="button"
      primary={primary}
      size={size}
      type={type}
      onClick={onClick}
    >
      {renderIcon()}
      {children}
    </StyleButton>
  )
}
