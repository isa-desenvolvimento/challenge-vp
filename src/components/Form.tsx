import { StyleFormContainer } from '@/styles/Form'
import { ItemProps } from '@/types/global'
import { Formik } from 'formik'
import React, { EventHandler } from 'react'
import { InputComponent } from '@/components/InputComponent'

export const Form = () => {
  const onSubmit = (data: ItemProps) => {
    console.log(data)
  }

  const renderForm = () => {
    return (
      <Formik
        onSubmit={(values) => onSubmit(values)}
        initialProps={{
          id: '',
          title: '',
          description: '',
          date: '',
          amount: '',
          status: '',
        }}
      >
        {(handleSubmit) => (
          <form onSubmit={handleSubmit}>
            <InputComponent placeholder="R$ 0.0" primary name="title" />
            <InputComponent placeholder="R$ 0.0" primary name="description" />
            <InputComponent placeholder="R$ 0.0" primary name="date" />
            <InputComponent placeholder="R$ 0.0" primary name="amount" />
            <InputComponent placeholder="R$ 0.0" primary name="status" />
          </form>
        )}
      </Formik>
    )
  }

  return <StyleFormContainer>{renderForm()}</StyleFormContainer>
}
