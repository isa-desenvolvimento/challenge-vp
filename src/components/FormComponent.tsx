import { StyleFormContainer } from '@/styles/Form'
import { FormikValues } from 'formik'
import { InputComponent } from '@/components/InputComponent'
import { BaseForm } from './BaseForm'

export const FormComponent = () => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const onSubmit = (data: FormikValues, action: any) => {
    console.log(data, action)
  }

  const renderBaseForm = () => (
    <StyleFormContainer role="form-component">
      <InputComponent
        placeholder="R$ 0.0"
        name="title"
        label="Formulário genérico"
      />
      <InputComponent
        placeholder="R$ 0.0"
        name="description"
        label="Formulário genérico"
      />
      <InputComponent
        placeholder="R$ 0.0"
        name="date"
        label="Formulário genérico"
      />
    </StyleFormContainer>
  )

  return (
    <BaseForm
      baseForm={renderBaseForm()}
      onSubmit={onSubmit}
      initialValues={{
        title: '',
        description: '',
        date: '',
        amount: '',
        status: '',
      }}
    />
  )
}
