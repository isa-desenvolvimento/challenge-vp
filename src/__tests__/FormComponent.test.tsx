import { renderWithProvider } from '../utils/helpers'
import { fireEvent, screen } from '@testing-library/react'

import { FormComponent } from '@/components/FormComponent'

describe('<FormComponent />', () => {
  beforeEach(() => renderWithProvider(<FormComponent />))

  it('should render FormComponent', () => {
    expect(screen.getByRole(/form-component/i).textContent).toContain(
      'Formulário genérico'
    )
  })

  it('should render input', () => {
    const input = screen.getByRole(/description/i)
    fireEvent.change(input, { target: { value: 23 } })
    expect(input.value).toBe('23')
  })

  // it('should render the styled-components tag-status', () => {
  //   expect(screen.getByRole(/tag-status/i)).toHaveStyle({
  //     background: '#36a4a8',
  //     color: '#ffff',
  //     'font-size': '1em',
  //     margin: '1em',
  //     border: 'none',
  //     'border-radius': '4px',
  //     padding: '0.8em 2em',
  //     width: '10%',
  //     'text-transform': 'uppercase',
  //     'text-overflow': 'ellipsis',
  //     cursor: 'pointer',
  //   })
  // })

  it('should render snapshot', () => {
    expect(screen.getByRole(/form-component/i)).toMatchSnapshot()
  })
})
