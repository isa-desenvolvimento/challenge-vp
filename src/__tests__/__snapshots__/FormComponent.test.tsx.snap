// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormComponent /> should render snapshot 1`] = `
.c0 {
  display: content;
  gap: 0.4rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 3rem;
  text-align: start;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  gap: 1rem;
  padding: 1rem;
  text-overflow: ellipsis;
  background: #eceff5;
  color: #3AA3AA;
  border: none;
  border-radius: 4px;
  width: auto;
  margin: 1rem;
}

.c3 {
  padding: 0;
  margin: 0;
  color: #3AA3AA;
  background: none;
  border: none;
  text-overflow: ellipsis;
}

.c1 {
  color: #60647C;
  text-overflow: ellipsis;
}

@media screen and (max-width:600px) {
  .c0 {
    display: revert;
    padding: 1rem;
  }
}

<div
  class="c0"
  role="form-component"
>
  <label
    class="InputComponent__StyleContainer-sc-o1q8z5-0"
    role="input-component"
  >
    <label
      class="c1"
    >
      Formulário genérico
    </label>
    <div
      class="c2"
      name="title"
    >
      <input
        class="c3"
        name="title"
        placeholder="R$ 0.0"
        role="title"
        value=""
      />
    </div>
  </label>
  <label
    class="InputComponent__StyleContainer-sc-o1q8z5-0"
    role="input-component"
  >
    <label
      class="c1"
    >
      Formulário genérico
    </label>
    <div
      class="c2"
      name="description"
    >
      <input
        class="c3"
        name="description"
        placeholder="R$ 0.0"
        role="description"
        value=""
      />
    </div>
  </label>
  <label
    class="InputComponent__StyleContainer-sc-o1q8z5-0"
    role="input-component"
  >
    <label
      class="c1"
    >
      Formulário genérico
    </label>
    <div
      class="c2"
      name="date"
    >
      <input
        class="c3"
        name="date"
        placeholder="R$ 0.0"
        role="date"
        value=""
      />
    </div>
  </label>
</div>
`;
